### PR TITLE
Canvas: Add "infinite" pan / zoom functionality

### LINKED ISSUE #77087: Canvas: Make extents grow with pan
Make canvas extents grow with pan. With [current approach for pan/zoom](https://github.com/grafana/grafana/pull/76705), panning to the canvas edges, causes pan to stop and "bounce" back. Elements are also cut off.

Image

Some considerations:
1. How does this impact the layout constraint system?
2. Does this become an option, and if so, where does that live?

To implement, inside of [app/features/canvas/runtime/SceneTransformationWrapper.tsx](https://github.com/grafana/grafana/blob/b0130ecb829925a3343abc208395a3bfdc1b9187/public/app/features/canvas/runtime/SceneTransformWrapper.tsx#L65), inside of the `<TransformWrapper/>` add something similar to this:  

        onPanning={(r, e) => {

        const mouseEvent = e as MouseEvent;
        // Get deltaX and deltaY from pan event and add it to current canvas dimensions
        let deltaX = mouseEvent.movementX;
        let deltaY = mouseEvent.movementY;
        if (deltaX > 0) {
          deltaX = 0;
        }
        if (deltaY > 0) {
          deltaY = 0;
        }
        scene.updateSize(scene.width - deltaX, scene.height - deltaY);
        scene.panel.forceUpdate();

      }}
        limitToBounds={false}

### PR DESCRIPTION
**Pull request by nmarrs**
Canvas pan / zoom has been a little limited by the inability to zoom out / pan over to elements that were not rendered due to the panel's size.

With the newly added `Infinite panning` editor option, users can now view and navigate their expansive canvases

[Demo dashboard](https://github.com/grafana/grafana/files/14719002/Airbus.Canvas.Demo-1711091923826.json)

Before

https://github.com/grafana/grafana/assets/22381771/416233f3-2455-4025-8adc-ce0d0ad80e92

After

https://github.com/grafana/grafana/assets/22381771/181ed4d9-a254-4859-9654-e8fd728e1ec3

Note, this is an _experimental_ feature that is behind a panel option that is behind the canvas panZoom feature toggle / functionality. I think it is worthwhile to get this out in the wild and get feedback from customers / users that have been requesting this functionality.

As it is experimental, there are a lot of rough edges, including:
- Only really supports elements that are top / left constrained (all others will be whacky given this solution actively resized the canvas which other constraints are reliant on)
- The canvas currently grows just to the bottom / right, meaning that although it looks like usable space, everything to the top left is not actually useable canvas space 🥴 
- TODO make an GH issues for the following:
- https://github.com/grafana/grafana/issues/85037
- Arrow connected to root can expand forever

https://github.com/grafana/grafana/assets/22381771/f687aaaa-3cd4-4f3f-a212-d501f7fa433e

Fixes https://github.com/grafana/grafana/issues/77087
Fixes https://github.com/grafana/support-escalations/issues/8962

### CODE REVIEW

File: public/app/plugins/panel/canvas/module.tsx [hunk 1]
```diff
@@ -39,6 +39,14 @@ export const addStandardCanvasEditorOptions = (builder: PanelOptionsEditorBuilde
     editor: PanZoomHelp,
     showIf: (opts) => config.featureToggles.canvasPanelPanZoom && opts.panZoom,
   });
+  builder.addBooleanSwitch({
+    path: 'infinitePan',
+    name: 'Infinite panning',
```
Comment (nmarrs) [inline public/app/plugins/panel/canvas/module.tsx:44]: Not in love with this naming, open to suggestions 😅

File: public/app/features/canvas/runtime/SceneTransformWrapper.tsx [hunk 2]
```diff
@@ -14,6 +14,15 @@ export const SceneTransformWrapper = ({ scene, children: sceneDiv }: SceneTransf
   const onZoom = (zoomPanPinchRef: ReactZoomPanPinchRef) => {
     const scale = zoomPanPinchRef.state.scale;
     scene.scale = scale;
+
+    if (scene.shouldInfinitePan) {
+      const isScaleZoomedOut = scale < 1;
+
+      if (isScaleZoomedOut) {
+        scene.updateSize(scene.width / scale, scene.height / scale);
```
Comment (nmarrs) [inline public/app/features/canvas/runtime/SceneTransformWrapper.tsx:22]: This logic could probably be improved

**Comment by Develer** [comment]
Hey @nmarrs, great work on pan and zoom! I like it! 👍🏻

Have found a couple of odd behaviour for follow-up tasks.

After zoom-out elements can disappear at all:

https://github.com/grafana/grafana/assets/1418924/d1037e21-864a-4a60-8497-23c8cf63fa84

**Comment by Develer** [comment]
Also found that for old canvas panels it's still cutting elements. Is there something I missed? Migration, etc.?

https://github.com/grafana/grafana/assets/1418924/75b4fcfa-7d36-46c1-9846-acd1fe9eb52b

**Comment by nmarrs** [comment]
> Also found that for old canvas panels it's still cutting elements. Is there something I missed? Migration, etc.?
> 
> 
> 
> https://github.com/grafana/grafana/assets/1418924/75b4fcfa-7d36-46c1-9846-acd1fe9eb52b
> 
> 
> 
> 

There is an additional editor option you need to toggle on underneath the pan zoom instructions. This option is only visible when pan / zoom is enabled and is marked as an experimental feature with limitations 😅

File: public/app/plugins/panel/canvas/module.tsx (hunk 1 above)
Comment (baldm0mma) [reply public/app/plugins/panel/canvas/module.tsx:44]: I actually think it's pretty declarative and self-explanatory. 👍

**Comment by Develer** [comment]
> > Also found that for old canvas panels it's still cutting elements. Is there something I missed? Migration, etc.?
> > 
> >   
> >     
> >     
> > 
> >     Screen.Recording.2024-03-22.at.16.34.41.mov
> 
> There is an additional editor option you need to toggle on underneath the pan zoom toggle. This option is only visible when pan / zoom is enabled and is marked as an experimental feature with limitations 😅

Oh, I see, now it works! Thanks!

Hm, I see some invisible border though 😁 Also for later

https://github.com/grafana/grafana/assets/1418924/3f9ffca3-f0d2-4499-9041-3f301004b800

File: public/app/plugins/panel/canvas/module.tsx (hunk 1 above)
Comment (Develer) [reply public/app/plugins/panel/canvas/module.tsx:44]: `addToggle` may be a good option, shorter, cleaner and also self-explanatory 😁

File: public/app/features/canvas/runtime/SceneTransformWrapper.tsx [hunk 3]
```diff
@@ -60,16 +88,21 @@ export const SceneTransformWrapper = ({ scene, children: sceneDiv }: SceneTransf
     }
   };
 
+  // Set panel content overflow to hidden to prevent canvas content from overflowing
+  scene.div?.parentElement?.parentElement?.parentElement?.parentElement?.setAttribute('style', `overflow: hidden`);
```
Comment (Develer) [inline public/app/features/canvas/runtime/SceneTransformWrapper.tsx:92]: Maybe we can use some `ref` variable?

**Review by baldm0mma — APPROVED** [review]
Worked as expected!

**Review by codeincarnate — APPROVED** [review]
Looks good to me! While there may still be some issues would make sense to get this out sooner rather than later 😄

**Review by Develer — APPROVED** [review]

**Pull request merged**